src/coreclr/jit/morph.cpp

```diff
@@ -13614,53 +13642,57 @@ void Compiler::fgMorphBlocks()
         lvSetMinOptsDoNotEnreg();
     }
 
-    /*-------------------------------------------------------------------------
-     * Process all basic blocks in the function
-     */
-
-    BasicBlock* block = fgFirstBB;
-    noway_assert(block);
-
-    do
+    // Morph all blocks. If we aren't then we just morph in normal bbNext order.
+    //
+    if (!optLocalAssertionProp)
     {
-#ifdef DEBUG
-        if (verbose)
+        // Note morph can add blocks downstream from the current block,
+        // and alter (but not null out) the current block's bbNext;
+        // this iterator ensures they all get visited.
+        //
+        for (BasicBlock* block : Blocks())
         {
-            printf("\nMorphing " FMT_BB " of '%s'\n", block->bbNum, info.compFullName);
+            fgMorphBlock(block);
         }
-#endif
+    }
+    else
+    {
+        // We are optimizing. Process in RPO.
+        //
+        fgRenumberBlocks();
+        EnsureBasicBlockEpoch();
+        fgComputeEnterBlocksSet();
+        fgDfsReversePostorder();
 
-        if (optLocalAssertionProp)
+        // Morph can introduce new blocks, so enable tracking block creation
+        //
+        unsigned const bbNumMax = fgBBNumMax;
+        fgEnableNewBlockTracking();
+        for (unsigned i = 1; i <= bbNumMax; i++)
         {
-            //
-            // Clear out any currently recorded assertion candidates
-            // before processing each basic block,
-            // also we must  handle QMARK-COLON specially
-            //
-            optAssertionReset(0);
+            BasicBlock* const block = fgBBReversePostorder[i];
+            fgMorphBlock(block);
         }
+        fgDisableNewBlockTracking();
 
-        // Make the current basic block address available globally.
-        compCurBB = block;
-
-        // Process all statement trees in the basic block.
-        fgMorphStmts(block);
-
-        // Do we need to merge the result of this block into a single return block?
-        if ((block->bbJumpKind == BBJ_RETURN) && ((block->bbFlags & BBF_HAS_JMP) == 0))
+        // If morph created new blocks, then morph them as well.
+        // Temporarily forbid creating even more new blocks so
+        // we don't have to iterate until closure.
+        //
+        // (TODO: verify that this potential out of order processing does not cause issues)
+        //
+        if (fgNewBBs->size() > 0)
         {
-            if ((genReturnBB != nullptr) && (genReturnBB != block))
+            JITDUMP("Morph created %u new blocks, processing them out of RPO\n", fgNewBBs->size());
+
+            INDEBUG(fgSafeBasicBlockCreation = false);
+            for (BasicBlock* const block : *fgNewBBs)
             {
-                fgMergeBlockReturn(block);
+                fgMorphBlock(block);
             }
```

Comment:
Is this a significant chunk of throughput cost? What is the improvement of this tracking mechanism alone?